pkg/sentry/kernel/msgqueue/msgqueue.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package msgqueue implements System V message queues.
+package msgqueue
+
+import (
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/log"
+	"gvisor.dev/gvisor/pkg/sentry/fs"
+	"gvisor.dev/gvisor/pkg/sentry/ipcutil"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+	ktime "gvisor.dev/gvisor/pkg/sentry/kernel/time"
+	"gvisor.dev/gvisor/pkg/syserror"
+	"gvisor.dev/gvisor/pkg/waiter"
+)
+
+const (
+	// System-wide limit for maximum number of queues.
+	maxQueues = linux.MSGMNI
+
+	// Maximum size of a queue in bytes.
+	maxQueueBytes = linux.MSGMNB
+
+	// Maximum size of a message in bytes.
+	maxMessageBytes = linux.MSGMAX
+)
+
+// Key is a user-provided identifier for message queues.
+type Key int32
+
+// ID is a kernel identifier for message queues.
+type ID int32
+
+// Registry contains a set of message queues that can be referenced using keys
+// or IDs.
+//
+// +stateify savable
+type Registry struct {
```

Comment:
If you define an IPC object as above, you can also write a common `Registry` type for all SysV IPC mechanisms. Replace `queues map[ID]*Queue` with `objs map[ID]*ipc.Object`, and have the new `ipc.Object` type have a interface field that can be used by the specific mechanisms to find their private datastructures. Use an interface to have the mechanisms provide their specific logic for registry operations:

```
// ipc_namespace.go
package kernel

type IPCRegsitryImpl interface {
    Create(ipc.ID) (*ipc.Object, error)
    ....
}

type Registry struct {
    userNS *auth.UserNamespace
    keysToIDs map[Key]ipc.ID
    lastIDUsed ipc.ID

    objs map[ipc.ID]*ipc.Object

    impl IPCRegistryImpl
}

func (r *Registry) FindByID(id ipc.ID) (*ipc.Object, error) {
    // Take locks, update ref counts, etc
    return r.objs[id], nil
}
```

```
package ipc

type ID int32

type Object struct {
    impl interface{}
}
```

```
package msgqueue

type Queue struct {
    o ipc.Object
    ...
}

func newQueue(...) *Queue {
    q := &Queue{...}
    q.o.impl = q
    return q
}

type registry struct {} // Implements kernel.IPCRegistryImpl

func (r *registry) Create(id ipc.ID) (*ipc.Object, error) {
    // Creation logic from FindOrCreate specific to msgqueues...
    ...
    q := newQueue(id)
    return &q.o
}
```

```
// sys_msgqueue.go
package linux

func findQueue(t *kernel.Task, id ipc.ID) (*msgqueue.Queue, error) {
    r := t.IPCNamespace().MsgqueueRegistry()
    qObj := r.FindByID(id)
    q, ok := q.Obj.(*msgqueue.Queue)
    if q == nil || !ok {
        // Non-msgqueue object in msgqueue registry?
        return nil, syserror.EINVAL
    }
    return q, nil
}
```

Reply:
As mentioned above, I added `ipc.Object` to package `ipc`. I'm not sure that `Registry` can be as easily abstracted, because each one has some small diffs. For example, `Semaphore.Registry` keeps track of indexes, which is not useful any where else. Similarly, func defs differ a bit from one registry to another. `Shm.Registry.FindOrCreate` takes pid and memory page size, `Sem.Registry.FindOrCreate` takes semaphore count, `Msgqueue.Registry.FindOrCreate` takes neither.

One thing we can do is to create `ipc.Registry` similar to `ipc.Object`, with fields that are common to all registries, and define some utility functions on it. We can then add it as a field to other registries (or maybe embed it?) and have them use it for common operations. That way we keep diff func defs with change, and avoid code repetition for things like allocating a new id.

Reply:
I've pushed a new patch with the updates we discussed. I've added an implementation of `ipc.Registry` in 4721df6, that works similar to `ipc.Object`, and is used by registries to define common funcs and fields.

The main diff between the implemented approach and the proposition is that instead of having `ipc.Registry` hold an implementation, we have implementation hold an `ipc.Registry` and use it for common operations. As I mentioned above, this lets us have diff func defs while avoiding repetition. Please take a look and let me know what you think.

Reply:
This looks great! A few minor suggestions around organizing common code below.

Reply:
Thank you! I uploaded a new patch addressing the comments. Please take a look and let me know if you have other comments / suggestions.